app/models/run.rb

```diff
@@ -220,28 +216,42 @@ def possible_timesave(timing)
 
   private
 
-  def stats_select_query(timing)
-    case timing
-    when Run::REAL
-      'segment_id,
-      STDDEV_POP(segment_histories.realtime_duration_ms) AS standard_deviation,
-      AVG(segment_histories.realtime_duration_ms) AS mean,
-      PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS median,
-      PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile10,
-      PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile90,
-      PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile99
-      '.squish
-    when Run::GAME
-      'segment_id,
-      STDDEV_POP(segment_histories.gametime_duration_ms) AS standard_deviation,
-      AVG(segment_histories.gametime_duration_ms) AS mean,
-      PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS median,
-      PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS percentile10,
-      PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS percentile90,
-      PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS percentile99
-      '.squish
-    else
-      raise 'Unsupported timing'
-    end
+  def segment_history_stats_query(timing)
+    timing_string = case timing
+                    when Run::REAL
+                      'real'
+                    when Run::GAME
+                      'game'
+                    else
+                      raise 'Unsupported timing'
+                    end
+
+    sql = %Q{
+      SELECT
+        segment_id,
+        STDDEV_POP(segment_histories.#{timing_string}time_duration_ms) AS standard_deviation,
+        AVG(segment_histories.#{timing_string}time_duration_ms) AS mean,
+        PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS median,
+        PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS percentile10,
+        PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS percentile90,
+        PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS percentile99 FROM segment_histories
+      INNER JOIN segments ON segments.id = segment_histories.segment_id
+      INNER JOIN runs ON runs.id = segments.run_id
+      LEFT JOIN (
+        SELECT segment_histories.id AS id, segment_histories.attempt_number AS attempt_number, segments.segment_number as segment_number
+        FROM segment_histories
+        INNER JOIN segments ON segments.id = segment_histories.segment_id
+        WHERE segments.run_id = #{id}
```

Comment:
I get really really uncomfortable doing interpolation in a SQL string. Can we do constructed statements instead (`'where segments.run_id = :id', {id: id}`)?

Same for interpolation for `timing_string` in fact. I know in this method it's guaranteed to be safe because of the `case` above, but often code gets copied/pasted in smaller parts or changed around slowly over time without remembering how important the guards are, so I always prefer that it be made foolproof from the beginning.

Reply:
As a note, I'm pretty positive you cannot use constructed statements with the column name, which is why the old version was the same query twice.

Reply:
Yeah, you can't do that with column names, do I just duplicated the query like the old version.